models/DataObject/Concrete/Dao/InheritanceHelper.php

```diff
@@ -421,43 +421,70 @@ protected function buildTree($currentParentId, $fields = '', $parentIdGroups = n
 
         if (!$parentIdGroups) {
             $object = DataObject::getById($currentParentId);
-            if (isset($params['language'])) {
-                $query = "SELECT a.language as language, b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE o_path LIKE ' . $this->db->quote($object->getRealFullPath() . '/%')
-                    . ' HAVING `language` = "' . $params['language'] . '" OR ISNULL(`language`)'
-                    . ' ORDER BY LENGTH(o_path) ASC';
-            } else {
-                $query = "SELECT b.o_id AS id $fields, b.o_classId AS classId, b.o_parentId AS parentId FROM objects b LEFT JOIN " . $this->storetable . ' a ON b.o_id = a.' . $this->idField . ' WHERE o_path LIKE ' . $this->db->quote($object->getRealFullPath().'/%') . ' GROUP BY b.o_id ORDER BY LENGTH(o_path) ASC';
-            }
-            $queryCacheKey = 'tree_'.md5($query);
 
-            if (self::$useRuntimeCache) {
-                $parentIdGroups = self::$runtimeCache[$queryCacheKey] ?? null;
-            }
+            if($object->hasChildren()){
+                /**
+                 * get the object keys once and reuse it because we don't have a key on o_path and it can take quite long when you have a lot of data
+                 *
+                 */
+                $query = 'SELECT o_id FROM objects WHERE o_path LIKE ' .  $this->db->quote($object->getRealFullPath() . '/%');
 
-            if (!$parentIdGroups) {
-                $result = $this->db->fetchAll($query);
+                if (self::$useRuntimeCache) {
+                    $queryCacheKey = 'tree_ids_'.md5($query);
+                    if(is_null(self::$runtimeCache[$queryCacheKey])){
```

Comment:
check for index
![image](https://user-images.githubusercontent.com/5137917/76983275-6ebca600-693d-11ea-9300-9beeaa8fc694.png)